Use OSProfiler WSGI middleware with Jaeger backend

BenjaminLudwigSAP opened the pull request:
Add OSProfiler WSGI middleware to Pecan.
Please review [the Charts PR](https://github.com/sapcc/helm-charts/compare/octavia_tracing?expand=1) first.

We use `custom-requirements.txt` in other projects for CCloud-specific requirements, so I used it here as well.

I am not really sure about the use of the `context` parameter in the invocation of `init_from_conf`. Apparently it's a context supplied in calls to Jaeger backend, and in Cinder it is [a custom Object returned by a `get_admin_context` function](https://github.com/sapcc/cinder/blob/stable/queens-m3/cinder/context.py#L246-L251). I set it to `{}`, [like Keystone does](https://github.com/sapcc/keystone/blob/stable/rocky-m3/keystone/common/profiler.py#L34).

BenjaminLudwigSAP committed:
Register OSProfiler WSGI middleware with Pecan

notandy commented on octavia/api/app.py:
```diff
@@ -106,6 +111,21 @@ def _wrap_app(app):
             )
 
+    # OSProfiler tracing middleware
+    if CONF.profiler.enabled and profiler_initializer and profiler_opts and profiler_web:
```
`CONF.profiler.enabled` config option is lazy-registered by import `osprofiler.opts`, so this check order will cause an exception if the library is missing:

```python
Traceback (most recent call last):
  File "/Users/d072895/Library/Application Support/JetBrains/Toolbox/apps/PyCharm-P/ch-0/203.6682.179/PyCharm.app/Contents/plugins/python/helpers/pydev/pydevd.py", line 1485, in _exec
    func()
  File "octavia/cmd/api.py", line 33, in main
    app = api_app.setup_app(argv=sys.argv)
  File "octavia/api/app.py", line 77, in setup_app
    wsme=pecan_config.wsme
  File "/Users/d072895/.virtualenvs/stein_python2.7/lib/python2.7/site-packages/pecan/__init__.py", line 91, in make_app
    app = wrap_app(app)
  File "octavia/api/app.py", line 115, in _wrap_app
    if CONF.profiler.enabled and profiler_initializer and profiler_opts and profiler_web:
  File "/Users/d072895/.virtualenvs/stein_python2.7/lib/python2.7/site-packages/oslo_config/cfg.py", line 2187, in __getattr__
    raise NoSuchOptError(name)
oslo_config.cfg.NoSuchOptError: no such option profiler in group [DEFAULT]
python-BaseException
Exception KeyboardInterrupt in <module 'threading' from '/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.pyc'> ignored
```

BenjaminLudwigSAP committed:
Fix: Evaluating profiler config option first throws exception

When registering the OSProfiler middleware with Pecan, the profiler
config option is evaluated first, before the profiler-related imports
are checked:

if CONF.profiler.enabled and profiler_initializer and profiler_opts and profiler_web:

Since the config options are loaded by osprofiler.opts, this check fails
if the osprofiler module is not available.

BenjaminLudwigSAP committed:
Trace method invocations in neutron network driver
